File: frontend/src/types/spoolmandb.ts
/**
 * SpoolmanDB types - Community filament database
 */

export interface SpoolmanDBManufacturer {
  id: string
  name: string
  is_active: boolean
  created_at: string
  updated_at: string
  filament_count?: number
}

export interface SpoolmanDBFilament {
  id: string
  external_id: string
  manufacturer_id: string
  manufacturer_name: string
  name: string
  material: string
  density: number | null
  diameter: number
  weight: number
  spool_weight: number | null
  spool_type: string | null
  color_name: string | null
  color_hex: string | null
  extruder_temp: number | null
  bed_temp: number | null
  finish: string | null
  translucent: boolean
  glow: boolean
  pattern: string | null
  multi_color_direction: string | null
  color_hexes: string | null
  is_active: boolean
  created_at: string
  updated_at: string
}

export interface SpoolmanDBManufacturerListResponse {
  manufacturers: SpoolmanDBManufacturer[]
  total: number
}

export interface SpoolmanDBFilamentListResponse {
  filaments: SpoolmanDBFilament[]
  total: number
  page: number
  page_size: number
}

export interface SpoolmanDBFilamentListParams {
  page?: number
  page_size?: number
  manufacturer_id?: string
  manufacturer_name?: string
  material?: string
  search?: string
  diameter?: number
}

export interface SpoolmanDBStats {
  total_manufacturers: number
  total_filaments: number
  materials: string[]
  last_sync: string | null
}

export interface SpoolmanDBSyncResponse {
  success: boolean
  manufacturers_added: number
  manufacturers_updated: number
  filaments_added: number
  filaments_updated: number
  message: string
}

File: backend/app/schemas/spoolmandb.py
"""Pydantic schemas for SpoolmanDB data."""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict


class SpoolmanDBManufacturerBase(BaseModel):
    """Base schema for manufacturer."""

    name: str


class SpoolmanDBManufacturerResponse(SpoolmanDBManufacturerBase):
    """Response schema for manufacturer."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    is_active: bool
    created_at: datetime
    updated_at: datetime


class SpoolmanDBManufacturerWithCount(SpoolmanDBManufacturerResponse):
    """Manufacturer with filament count."""

    filament_count: int = 0


class SpoolmanDBFilamentBase(BaseModel):
    """Base schema for filament."""

    external_id: str
    name: str
    material: str
    density: float | None = None
    diameter: float
    weight: int
    spool_weight: int | None = None
    spool_type: str | None = None
    color_name: str | None = None
    color_hex: str | None = None
    extruder_temp: int | None = None
    bed_temp: int | None = None
    finish: str | None = None
    translucent: bool = False
    glow: bool = False
    pattern: str | None = None
    multi_color_direction: str | None = None
    color_hexes: str | None = None


class SpoolmanDBFilamentResponse(SpoolmanDBFilamentBase):
    """Response schema for filament."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    manufacturer_id: UUID
    is_active: bool
    created_at: datetime
    updated_at: datetime


class SpoolmanDBFilamentWithManufacturer(SpoolmanDBFilamentResponse):
    """Filament with manufacturer details."""

    manufacturer_name: str


class SpoolmanDBFilamentListResponse(BaseModel):
    """Paginated list of filaments."""

    filaments: list[SpoolmanDBFilamentWithManufacturer]
    total: int
    page: int
    page_size: int


class SpoolmanDBManufacturerListResponse(BaseModel):
    """List of manufacturers."""

    manufacturers: list[SpoolmanDBManufacturerWithCount]
    total: int


class SpoolmanDBSyncResponse(BaseModel):
    """Response from sync operation."""

    success: bool
    manufacturers_added: int
    manufacturers_updated: int
    filaments_added: int
    filaments_updated: int
    message: str


class SpoolmanDBStatsResponse(BaseModel):
    """Statistics about SpoolmanDB data."""

    total_manufacturers: int
    total_filaments: int
    materials: list[str]
    last_sync: datetime | None = None
